### DISCUSSION TITLE
Is it possible to detect supported browser or switch to turn off p2p when error

### DISCUSSION THREAD

**Discussion by ElonJTrump** [Novage/p2p-media-loader#423] [General]
I found an problem with some Safari browser on some Apple device. (The most case error in hls.js debug log is can't decode but when I'm not use P2P no problem with the same playlis) So I think is it possible to add 'isSupported' method and on error event listener to switch to normal HLS without P2P.

I try to add 'isP2PDisabled: true' when error happended but seem to not solve.

Can we share some idea for handle this situation ?

**Comment by mrlika** [comment]
There is [`isSupported`](https://github.com/video-dev/hls.js/blob/master/docs/API.md#getting-started) method in Hls.js that should do the checks.

**Comment by DimaDemchenko** [comment]
@stevenking888 In case Hls.isSupported() doesn't solve your problem, please provide the following details:
- hls.js Version:  
- Device Type & OS Version.
- P2P Configuration
- Stream Playlist

**Comment by xiaoya2025** [comment]
> I found an problem with some Safari browser on some Apple device. (The most case error in hls.js debug log is can't decode but when I'm not use P2P no problem with the same playlis) So I think is it possible to add 'isSupported' method and on error event listener to switch to normal HLS without P2P.
> 
> I try to add 'isP2PDisabled: true' when error happended but seem to not solve.
> 
> Can we share some idea for handle this situation ?

I use mac xcode Simulator test,even if ios version > 17.1.1,can not use ManagedMediaSource api play video,but  use Real machine debugger can play video,so i use js check it.

hls.js Hls.isSupported() Inaccurate judgment
dash.js dashjs.supportsMediaSource() Inaccurate judgment
shaka.Player.isBrowserSupported() Inaccurate judgment

Accurate judgment：

```

                  if (shaka.Player.isBrowserSupported()) {
                      let supportFlag = false;
                      const mediaType = 'video/mp4; codecs="mp4a.40.2,avc1.4d4015"';
                      if ("ManagedMediaSource" in window && ManagedMediaSource.isTypeSupported(mediaType)) supportFlag = true;
                      if ("MediaSource" in window && MediaSource.isTypeSupported(mediaType)) supportFlag = true;

                      if (!supportFlag) {
                          video.src = url;
                          return;
                      }
                      const player = new shaka.Player();
                      player.configure({
                          streaming: {
                              // bufferingGoal: 120
                          }
                      });
                     
                      await player.attach(video);
                      // Try to load a manifest.
                      // This is an asynchronous process.
                      try {
                          await player.load(url);
                      } catch (e) {
                          // onError is executed if the asynchronous load fails.
                          console.log(e);

                      }

                  } else if (video.canPlayType('application/vnd.apple.mpegurl')) {
                      video.src = url;
                  }
```
![test](https://github.com/user-attachments/assets/aa9e6749-fb79-4355-b8db-b70c450d4fab)

**Comment by mrlika** [comment]
There is a bug in the iPhone simulator. It doesn't support required MMS (it is how MSE is called on iOS) feature/codecs for Hls.js to run. Do test/development on iPad simulator or real iOS device.

**Comment by mrlika** [comment]
Xcode was recently updated, and we will test if the issue with MMS in the iPhone simulator is fixed there.

**Comment by mrlika** [comment]
MMS is fixed and works in iOS 18 simulators:

<img width="362" alt="image" src="https://github.com/user-attachments/assets/780ff963-ea1f-43d2-9902-72d06176bcaa">
<img width="363" alt="image" src="https://github.com/user-attachments/assets/b5a08f8c-b60b-4812-8175-6c6b543b6a5e">

**Comment by ElonJTrump** [comment]
I found problem only on Mac Safari. When play a while (not over 5 minute) double tested with multiple Mac devices

Mac pass these condtions

window.MediaSource !== 'undefined' 
window.WebKitMediaSource !== 'undefined';
RTCPeerConnection !== 'undefined'
webkitRTCPeerConnection !== 'undefined'
mozRTCPeerConnection !== 'undefined'

<img width="604" alt="Screenshot 2024-09-22 072258" src="https://github.com/user-attachments/assets/11e40638-869a-4151-abcc-2511b470a668">

**Comment by mrlika** [comment]
We can't reproduce the issue with Safari. All the videos play for us for over 5 minutes on Safari. Is the problem reproducible with our demo?

**Comment by ElonJTrump** [comment]
I think I found the root of problem on macOS. That is macOS disconnect websocket when play for a while. I use 'wt-tracker' as my tracker

<img width="734" alt="Screenshot 2024-10-01 163655" src="https://github.com/user-attachments/assets/7a334acd-1215-45e8-a628-a66445f789a6">
